### PR TITLE
Added result caching for QueryBuilder

### PR DESCRIPTION
**Pull request by twoleds**
|      Q       |   A
|------------- | -----------
| Type         | improvement
| Fixed issues | 

#### Summary

Added `enable/disableResultCache` method in QueryBuilder for result caching.

I use `doctrine/dbal` in several projects and the result caching is annoying. Only way I can use result caching is calling `executeCacheQuery` method in `Connection`. I cannot use `fetch*` methods in `QueryBuilder`.

I repeat this piece of code:
```php
$builder = $connection->createQueryBuilder();
// ....
$result = $connection->executeCacheQuery(
    $builder->getSQL(),
    $builder->getParameters(),
    [],
    new \Doctrine\DBAL\Cache\QueryCacheProfile($cacheLifetime)
)->fetchAllAssociative();
```

With this improvement it will be:
```php
$builder = $connection->createQueryBuilder();
// ....
$builder->enableResultCache(300)->fetchAllAssociative();
```

What do you think about it? Is it possible to implement something like that?

### CODE REVIEW

File: src/Connection.php [hunk 1]
```diff
@@ -526,9 +526,13 @@ public function setAutoCommit($autoCommit)
...
      * @throws Exception
      */
-    public function fetchAssociative(string $query, array $params = [], array $types = [])
-    {
-        return $this->executeQuery($query, $params, $types)->fetchAssociative();
+    public function fetchAssociative(
+        string $query,
+        array $params = [],
+        array $types = [],
+        ?QueryCacheProfile $qcp = null
```
Comment (morozov) [inline src/Connection.php]: Is this really necessary? There are already too many optional parameters in these methods.

The query builder should be able to call `$connection->executeQuery(..., $qcp)->fetchAssociative()` instead.

Comment (twoleds) [reply src/Connection.php]: You are right it is not necessary for `Connection` class, I use `QueryBuilder` 100% time I will update the code soon. Thanks for reply and good point.

Comment (twoleds) [reply src/Connection.php]: I have updated the code and now I will think about how to test it, whether to extend the existing fetch* tests or create a new set testing methods for caching. What do you think?

File: src/Query/QueryBuilder.php [hunk 2]
```diff
@@ -1564,4 +1577,52 @@ public function __clone()
...
+    {
+        return $this->resultCacheProfile;
+    }
+
+    /**
+     * Set a cache profile for the result caching.
+     *
+     * @return $this
+     */
+    public function setResultCacheProfile(?QueryCacheProfile $cacheProfile): self
```
Comment (morozov) [inline src/Query/QueryBuilder.php]: Having two ways to enable-disable caching looks like overkill. When a method just "sets" something, it looks creepy. The name of the method should convey some semantics. `enableResultCache()` and `disableResultCache()` look fine to me from the naming perspective.

How about having `enableResultCache()` accept an instance of `QueryCacheProfile`? That should be sufficient.

**Comment by morozov** [comment]
@twoleds please rebase your branch on top of `3.4.x`. The default `3.3.x` branch accepts only bug fixes.

File: src/Query/QueryBuilder.php (hunk 2 above)
Comment (twoleds) [reply src/Query/QueryBuilder.php]: I took inspiration from `doctrine/orm`, where something similar is implemented. There are a lot of methods for result caching: `setResultCacheProfile`, `setResultCache`,  `useResultCache`, `enableResultCache`, `disableResultCache`, `setResultCacheLifetime`,  `expireResultCache`, `getExpireResultCache`, `getQueryCacheProfile`, `setResultCacheId`, `setResultCacheDriver` (deprecated), `getResultCacheDriver` (deprecated), `getResultCacheLifetime` (deprecated), `getResultCacheId` (deprecated) in `AbstractQuery` class. Too much methods for the same thing. I also don't like many methods for the same thing and that's why I choose only 4. The `enableResultCache` where you can simply pass the lifetime and optionally cache key and it's not necessary create new instance of `QueryCacheProfile` every time and if someone need more control, then he can use `setQueryCacheProfile`. But I think that for many cases setting the lifetime will be sufficient. And I also wanted to keep the naming and functioning of the methods according to `doctrine/orm` package. If someone is used to using these methods in `doctrine/orm', it would be good to keep their functionality similar. 

https://github.com/doctrine/orm/blob/2.12.x/lib/Doctrine/ORM/AbstractQuery.php#L717

```php
// variant A
$builder->enableResultCache(300)->fetchAllAssociative();

// variant B
$builder->enableResultCache(new QueryCacheProfile(300))->fetchAllAssociative();
```

I like the variant A because of less typing and it is simillar with the method in `doctrine/orm`. What do you thing about it?

Comment (morozov) [reply src/Query/QueryBuilder.php]: The documentation suggests that the `QueryCacheProfile` instance needs to be created by the user and passed to the connection:
https://github.com/doctrine/dbal/blob/3ae5312bbd0b5fe271e90ca8d957038d14e1c840/docs/en/reference/caching.rst#L42-L43

The `QueryCacheProfile` class doesn't have mandatory parameters in the constructors and have some builder-like methods:
https://github.com/doctrine/dbal/blob/595e2df1590257bc5bb0423a1ba1f332e01be0f3/src/Cache/QueryCacheProfile.php#L28-L36 https://github.com/doctrine/dbal/blob/595e2df1590257bc5bb0423a1ba1f332e01be0f3/src/Cache/QueryCacheProfile.php#L90-L103

With that said, I believe it's the user who should be responsible for building a `QueryCacheProfile` object and passing it to the query builder. I don't think we need to introduce another way of building `QueryCacheProfile`. If there are any changes in the `QueryCacheProfile` constructor, they would have to be reflected in the currently proposed `enableResultCache()`. We should avoid that.

To sum this up: I'm fine with introducing three new methods:
1. `enableResultCache(QueryCacheProfile $cacheProfile)`,
2. `disableResultCache()`,
3. `getResultCacheProfile()`.

As for the ORM, as you said, it exposes too many methods for doing essentially the same. If you want to make the DBAL and the ORM APIs consistent, I believe we should model the ORM after the DBAL (which will have a new and well-though API), not introduce a new API to make it look like the existing bloated API.

Comment (twoleds) [reply src/Query/QueryBuilder.php]: I understand, I changed the code and I put some new tests there, can you please check it if it is enough thanks

File: src/Query/QueryBuilder.php [hunk 3]
```diff
@@ -123,6 +124,13 @@ class QueryBuilder
      */
     private int $boundCounter = 0;
 
+    /**
+     * The query cache profile used for caching results.
+     *
+     * @var QueryCacheProfile|null
+     */
+    private $resultCacheProfile = null;
```
Comment (derrabus) [inline src/Query/QueryBuilder.php]:
```suggestion
     */
    private ?QueryCacheProfile $resultCacheProfile = null;
```

Comment (twoleds) [reply src/Query/QueryBuilder.php]: changed, thanks for suggestion

**Review by morozov — DISMISSED** [review]
Looks good. Good work @twoleds! I'll let the PR hang for a bit more in case anybody has feedback.

**Review by derrabus — DISMISSED** [review]

**Review by greg0ire — COMMENTED** [review]
Is there documentation to update?

**Comment by morozov** [comment]
Yeah, a section on caching could added to this document: https://github.com/doctrine/dbal/blob/d6aaae9a963f6cd49f5c5899c2054e918e630c82/docs/en/reference/query-builder.rst#L1

**Comment by twoleds** [comment]
I provided some documentation for the query builder, but my English is not so good, please check it if I wrote it correct.

File: docs/en/reference/query-builder.rst [hunk 4]
```diff
@@ -369,3 +369,26 @@ in your query as a return value:
...
     ;
     // SELECT id, name FROM users WHERE email = ?
+
+Caching
+-------
+
+To use the result cache, it is necessary to call the method
+``enableResultCache($cacheProfile)`` and pass a instance of
+``Doctrine\DBAL\Cache\QueryCacheProfile`` with lifetime in seconds.
+Optionally can be specified cache key.
```
Comment (SenseException) [inline docs/en/reference/query-builder.rst]:
```suggestion
``Doctrine\DBAL\Cache\QueryCacheProfile`` with a cache lifetime value in seconds.
A  cache key can optionally be added if needed.
```

File: docs/en/reference/query-builder.rst [hunk 5]
```diff
@@ -369,3 +369,26 @@ in your query as a return value:
...
+    $queryBuilder
+        ->select('id', 'name')
+        ->from('users')
+        ->enableResultCache(new QueryCacheProfile(300, 'some-key'))
+    ;
+
+.. note::
+
+    For more details and how to configure the result cache check caching section in
+    documentation.
```
Comment (SenseException) [inline docs/en/reference/query-builder.rst]: Can you please link to the caching section of the docs?

**Review by SenseException — DISMISSED** [review]

File: docs/en/reference/query-builder.rst (hunk 4 above)
Comment (twoleds) [reply docs/en/reference/query-builder.rst]: Thanks for help, I updated it.

File: docs/en/reference/query-builder.rst (hunk 5 above)
Comment (twoleds) [reply docs/en/reference/query-builder.rst]: Can somebody help me how to add link there. I tried ``` `caching section in
    documentation <./caching.rst>` ``` and ``` `caching section in
    documentation <https://www.doctrine-project.org/projects/doctrine-dbal/en/latest/reference/caching.html>` ```. I don't like absolute path. It is posible use realtive link there? I don't know how it works. RST files are for me new thing.

File: src/Query/QueryBuilder.php [hunk 6]
```diff
@@ -1548,4 +1559,37 @@ public function __clone()
...
+    {
+        $this->resultCacheProfile = null;
+
+        return $this;
+    }
+
+    /**
+     * Gets the cache profile for the result caching.
+     */
+    public function getResultCacheProfile(): ?QueryCacheProfile
```
Comment (morozov) [inline src/Query/QueryBuilder.php]: I'm curious if we really need to introduce this method. The builder will use the result cache profile for executing the queries that it has built. Who else except for the builder is supposed to use it?

File: docs/en/reference/query-builder.rst (hunk 5 above)
Comment (morozov) [reply docs/en/reference/query-builder.rst]: From [StackOverflow](https://stackoverflow.com/a/66289376/146187):

> An existing file {example.rst} may be linked to with the following syntax:
> ```
> :ref:`Optional Link text <example>`
> ```
See the following bits for example: https://github.com/doctrine/dbal/blob/14675c0436fcd65b3727db74e691412ee4930622/docs/en/reference/data-retrieval-and-manipulation.rst#L344 https://github.com/doctrine/dbal/blob/7b24c55dc81231fb8265dcc0d0505c19e5753472/docs/en/reference/types.rst#L481

Comment (morozov) [reply docs/en/reference/query-builder.rst]: @SenseException I'd leave the exercise of linking documentation articles to the maintainers. It's not trivial and there's no way to check the correctness of their change for a first-time contributor.

File: src/Query/QueryBuilder.php (hunk 6 above)
Comment (derrabus) [reply src/Query/QueryBuilder.php]: That is a very good point. I think, we should remove the method.

Comment (twoleds) [reply src/Query/QueryBuilder.php]: The method was successfully eliminated.

File: docs/en/reference/query-builder.rst (hunk 5 above)
Comment (twoleds) [reply docs/en/reference/query-builder.rst]: I put ``` :ref:`Caching <caching>` ``` there, hopefully it will work. Thanks.

**Review by morozov — APPROVED** [review]

**Comment by morozov** [comment]
Thanks, @twoleds!

**Comment by twoleds** [comment]
Thanks for help to everyone, it was an interesting exprerience for me, I'm very happy that we have reached a successful end. Life will be a little bit easier. ❤️

**Comment by rvanlaak** [comment]
This feature is so cool, thank you for your contribution @twoleds ! 👏
